pkg/cgroups/cgroups.go

```diff
@@ -47,6 +52,71 @@ func GetCgroupRoot() string {
 	return cgroupRoot
 }
 
+// getCgroupRootForKind tries to determine a cgroup v2 root path for attaching
+// the BPF cgroup programs when Cilium is running on Kind. The determined path
+// is not used for mounting, and it is set to "cgroupRoot" after the mounting
+// has succeeded.
+//
+// When running on Kind, each k8s node runs in a separate non-host netns container
+// on the same host. This means, that there might be multiple cilium-agent
+// instances on the single host. When bpf_sock LB is enabled, each cilium-agent
+// would try to attach the program to the same cgroup v2 root. In such case,
+// the bpf_sock based LB would not work because the program would try to access
+// the same LB BPF map which is not shared among Kind nodes.
+//
+// To fix this, we need to attach the program to cgroup v2 root of each Kind node.
+// Pods on a Kind node belong to a cgroup v2 which name format is the following:
+//
+//		0::/system.slice/docker-$(SOME_NODE_ID).scope/kubelet/kubepods/burstable/pod-$(SOME_POD_ID)
+//
+// So if we attach the program to "$(CGROUP_MOUNT_DIR)/system.slice/docker-$(SOME_NODE_ID).scope/kubelet/",
+// it's guaranteed that it will be run for all pods on that node.
+//
+// However, there is one problem when net_cls or net_prio of the legacy cgroup (v1)
+// is enabled together with cgroup v2. If any of these controllers store class_id
+// or prioridx (which is the case for Docker) in the socket cgroup data struct,
+// the kernel won't be able to retrieve the exact root, and it will fallback to
+// the default hierarchy of cgroup v2 (aka the mount path) [1]. This renders the
+// programs attached in a sub-hierarchy ineffective.
+//
+// [1]: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/cgroup.h?h=v5.10#n837
+func getCgroupRootForKind() (string, error) {
+	f, err := os.Open("/proc/self/cgroup")
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	cgroupPath := ""
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.Contains(line, "net_cls") || strings.Contains(line, "net_prio") {
+			return "", fmt.Errorf("cgroups v1 net_cls or net_prio detected which interferes with cgroup v2")
+		}
+
+		if strings.HasPrefix(line, "0::") { // cgroup v2 starts with "0::"
+			re := regexp.MustCompile(`^.*kubelet`)
```

Comment:
nit: this can be moved outside the loop (i.e. no need to recompile the regex at each iteration)

Reply:
This will happen only once, as there will be at most one instance of `0::`.